src/client/Microsoft.Identity.Client/Internal/Requests/OnBehalfOfRequest.cs

```diff
@@ -127,30 +127,46 @@ protected override async Task<AuthenticationResult> ExecuteAsync(CancellationTok
-                _ccsRoutingHint = CoreHelpers.GetCcsClientInfoHint(clientInfo.UniqueObjectIdentifier,
-                                                                                  clientInfo.UniqueTenantIdentifier);
+                AuthenticationRequestParameters.RequestContext.Logger.Info("[OBO request] Long-running OBO flow, trying to refresh using an refresh token flow.");
 
-                var msalTokenResponse = await SilentRequestHelper.RefreshAccessTokenAsync(cachedRefreshToken, this, AuthenticationRequestParameters, cancellationToken)
-                .ConfigureAwait(false);
+                // Look for a refresh token
+                MsalRefreshTokenCacheItem cachedRefreshToken = await CacheManager.FindRefreshTokenAsync().ConfigureAwait(false);
 
-                return await CacheTokenResponseAndCreateAuthenticationResultAsync(msalTokenResponse).ConfigureAwait(false);
-            }
+                // If a refresh token is not found, fetch a new access token
+                if (cachedRefreshToken != null)
+                {
+                    var clientInfo = ClientInfo.CreateFromJson(cachedRefreshToken.RawClientInfo);
+                    _ccsRoutingHint = CoreHelpers.GetCcsClientInfoHint(clientInfo.UniqueObjectIdentifier,
+                                                                                      clientInfo.UniqueTenantIdentifier);
+
+                    var msalTokenResponse = await SilentRequestHelper.RefreshAccessTokenAsync(cachedRefreshToken, this, AuthenticationRequestParameters, cancellationToken)
+                    .ConfigureAwait(false);
 
-            AuthenticationRequestParameters.RequestContext.Logger.Info("[OBO request] No Refresh Token was found in the cache. Fetching OBO token from ESTS");
+                    return await CacheTokenResponseAndCreateAuthenticationResultAsync(msalTokenResponse).ConfigureAwait(false);
+                }
+
+                AuthenticationRequestParameters.RequestContext.Logger.Info("[OBO request] No Refresh Token was found in the cache. Fetching OBO token from ESTS");
+            }
+            else
+            {
+                AuthenticationRequestParameters.RequestContext.Logger.Info("[OBO request] Normal OBO flow, skipping to fetching access token via OBO flow.");
+            }
 
             return await FetchNewAccessTokenAsync(cancellationToken).ConfigureAwait(false);
         }
 
         private async Task<AuthenticationResult> FetchNewAccessTokenAsync(CancellationToken cancellationToken)
         {
             var msalTokenResponse = await SendTokenRequestAsync(GetBodyParameters(), cancellationToken).ConfigureAwait(false);
+            
+            // We always retrieve a refresh token in OBO but we don't want to cache it for normal OBO flow, only for long-running OBO
+            if (!IsLongRunningObo())
+            {
+                msalTokenResponse.RefreshToken = null;

```

Comment:
For OBO, AAD returns a refresh token even if not sending `offline_access` scope. So solution is to set it to null after receiving the response.